### ISSUE TITLE
[RFC] Proposed implementation for Row and Cell editing

### ISSUE THREAD

**Issue by oliviertassinari** [mui-org/material-ui-x#1032]
Related issues: #203, #204.

TODO:

- [ ] Look into mobile UX
- [ ] Structure of the docs

### Batching

I propose we don't support batching updates. It seems to be a trend from an older decade. Nowadays, end-users expert more and more live updates.

### Row vs. Cell editing

Developers have upvoted the row editing support more than the cell one. I'm not too surprised as the component is meant to handle tabular data, often stored in a relational database, where records are stored in rows (e.g. Postgres/MySQL). 

We need to support both. I propose we start with implement Cell editing first, then move to Row editing as a different mode that batches the cell editing of the whole row.

Cell editing is likely to be implemented with a row menu as documented in #517. This is important for supporting a high number of columns. The row menu needs to include multiple actions when inactive: "delete", "edit" and "save", "cancel" once active.

### Enabling the feature

We make the feature opt-in. There are different ways to enable the feature. It can either be enabled for all the cells, per column or per cell. 

### Editors

We need a different editor per type of column. I propose we support the default types and allow developers to provide a custom cell editor. Note that we don't have an `InputNumber` component yet: https://github.com/mui-org/material-ui/issues/19154.  

We can add a `renderEditCell` prop, matching the `renderCell` approach in the column definitions.

```jsx
  renderEditCell = {
    row,
    field: column.field,
    value: column.valueGetter(row[column.field]),
    onChange, // when value is valid and changed, to be picked up at the next "Stop edition" event, can come from outside the cell.
    onChangeCommitted, // when developers want to stop edition. Parsed value.
  };
```

https://next.material-ui.com/api/slider-unstyled/

### Start editing

The different ways to trigger the edition. I think that we need to document them:

- Edit key:
  - <kbd>Enter</kbd>: focus the field at the end.
  - <kbd>Delete</kbd>: empty the field (do not focus it)
  - <kbd>Backspace</kbd>: empty the field (do not focus it)
- Printable key: focus the field, empty the field and start typing the letter into it
- Double click
- Imperative API call: `toggleCellMode(mode: 'edit' | 'view')`,

### During edition

End-users can revert changes with <kbd>Ctrl</kbd>+<kbd>Z</kbd> and reply changes with <kbd>Shift</kbd>+<kbd>Ctrl</kbd>+<kbd>Z</kbd>.

### Stop editing

The edition mode can be stopped by:

- Focusing another cell: current changes are retained
- Pressing <kbd>Enter</kbd>: it also moves the focus to the next cell (vertically), in edition mode.
- Pressing <kbd>Escape</kbd>: it restores the previous state
- Pressing <kbd>Tab</kbd>: behaves as <kbd>Enter</kbd> but move to the next cell horizontally.
- Pressing <kbd>Shift</kbd>+<kbd>Tab</kbd>: behaves as <kbd>Tab</kbd> but move to the previous cell horizontally.
- Imperative API call: `toggleCellMode(mode: 'edit' | 'view')`,

Clicking outside of the data grid doesn't stop editing.

### After edition

End-users can revert changes with <kbd>Ctrl</kbd>+<kbd>Z</kbd> and reply changes with <kbd>Shift</kbd>+<kbd>Ctrl</kbd>+<kbd>Z</kbd>.

### Server-side callback

Developers need to be able to store the edits back into the database. Maybe we could add a `onRowChange` callback to handle cell updates, row updates, row deletion, row addition.

This also opens a new discussion about `rows` being a controllable prop or not. If the data grid is able to modify the rows without asking for permission from the parent, then it's an uncontrollable prop, and might be better named: `defaultRow`.

### Add row button

It could be interesting to add a "Add row" button in the toolbar to add a new row at the top of the data grid, starting in edition mode.

### External form

I don't think that we need to do anything in order to support this mode, however, adding a demo in the documentation to reproduce it would be great. For instance:

- https://www.telerik.com/kendo-react-ui/components/grid/editing/editing-external-form/
- https://editor.datatables.net/

### Docs

cc @mui-org/x

**Comment by mbrookes** [comment]
For the row menu, I'd be keen to see an overlay toolbar, such as that in Gmail:

![image](https://user-images.githubusercontent.com/357702/107705833-45c0a480-6cb7-11eb-8bd5-d1d87428fea4.png)

**Comment by oliviertassinari** [comment]
@mbrookes So a preference for the last benchmark in #517

**Comment by mbrookes** [comment]
Yes, or the first if there are too many options, and it needs to overflow into a dropdown.

(Feel free to hide this discussion since there's a issue dedicated to it.)

**Comment by dtassone** [comment]
Things to consider:

- The feature should be flexible enough to turn a cell into edit mode, through an event such as double click or the click of a button.
- Basic UX should be mobile friendly
- A cell should be editable according to its value, it has to be dynamic. Ie ( row.status === draft ?  ) Or maybe if the user has the role...
- The new value of a cell can be validated: Ie email format? Or even username exist (DB call)
- render different input component according to the type of value to edit

Pre-req
- fix issue with number as IDs (Events get the row id using the data-id attr stored in the dom as string, if original type is number then we won't retrieve the corresponding row)

Out of scope:
External Form: Yes, having an overlay with a form can be done outside the scope of edit rows. 
Providing a row menu or an overlay toolbar should be treated as a separate issue.
Providing a delete row button should be considered separately
Providing a feature to add a row can also be considered in a separate issue.

**Comment by dtassone** [comment]
API: 

- Similar to `renderCell` in `ColDef`, we can consider adding a `renderEditCell` that would allow users to render any component to edit this particular column
- In order to turn a cell in edit mode, we can add an `editCell` method in `apiRef` that allows to turn the cell in edit mode and call `renderEditCell`
- In order to dynamically enable/disable the edit of cell, we add an `isCellEditable` method in `apiRef` that check if a cell is editable and return a boolean. That would allows to enable or disable an edit button or menu...
- In order to validate the new cell value,  we can add a `onCellValueChange` method that would be called when the user has finished editing the cell (Pressing enter or ok click...)

**Comment by oliviertassinari** [comment]
> The feature should be flexible enough to turn a cell into edit mode, through an event such as double click or the click of a button.

An imperative method in API sounds good enough. *Updated*

> Basic UX should be mobile friendly

I haven't looked into this aspect, more research would be great. I have added it to a TODO item.

> A cell should be editable according to its value, it has to be dynamic. Ie ( row.status === draft ? ) Or maybe if the user has the role...

True, I think that a `renderEditCell` like method in the column definition could be doing it. *Updated*

> The new value of a cell can be validated: Ie email format? Or even username exist (DB call)

I don't envision special treatments needed. I would imagine that email validation is done by customizing the `renderEditCell`, providing error feedback when the value is wrong, as well as customizing `valueParser` or valueSetter` to not set any new value if not correct.

> render different input component according to the type of value to edit

Covered under the "Editors" section.

> Out of scope:

What do you mean by out of scope? If you mean that they aren't as high priority and should be solved after, then we agree.

**Comment by dtassone** [comment]
> > The feature should be flexible enough to turn a cell into edit mode, through an event such as double click or the click of a button.
> 
> An imperative method in API sounds good enough. _Updated_
> 
> > Basic UX should be mobile friendly
> 
> I haven't looked into this aspect, more research would be great. I have added it to a TODO item.
> 
> > A cell should be editable according to its value, it has to be dynamic. Ie ( row.status === draft ? ) Or maybe if the user has the role...
> 
> True, I think that a `renderEditCell` like method in the column definition could be doing it. _Updated_

I don't think that will be enough. `renderEditCell` will be called when the cell is already in edit mode. Post Edit, chain of events. We need a pre-edit event in order to enable or disable an edit button for example.
> 
> > The new value of a cell can be validated: Ie email format? Or even username exist (DB call)
> 
> I don't envision special treatments needed. I would imagine that email validation is done by customizing the `renderEditCell`, providing error feedback when the value is wrong, as well as customizing `valueParser` or `valueSetter` to not set any new value if not correct.

Actually yes, validation can be done in the `renderEditCell`, and we can also consider that the row update should be done their as well. Therefore I do not see the point of `valueParser` or `valueSetter`

> 
> > render different input component according to the type of value to edit
> 
> Covered under the "Editors" section.
> 
> > Out of scope:
> 
> What do you mean by out of scope? If you mean that they aren't as high priority and should be solved after, then we agree.
Yes I suggest that we create separate issues for the row menu or overlay... These features can be treated separately and are not directly impacting the row edition. They will just improve the overall grid UX.
Adding or deleting a row can also be treated in a separate issue.

**Comment by oliviertassinari** [comment]
> we create separate issues for the row menu or overlay...

Do you mean #517? Why do I have to link it 3 times 😁

**Comment by oliviertassinari** [comment]
>  Therefore I do not see the point of `valueParser` or `valueSetter`

`valueParser` Is about converting the value rendered inside `renderEditCell` into a value that can be stored inside the row. `valueSetter` Is about storing the new value inside the row.

It seems that `renderEditCell` should be **stateless** in order to implement all the controls. By making it stateless, we make the format and update steps something that `renderEditCell` can't have full control over hence these two functions would still be required.

**Comment by dtassone** [comment]
> > we create separate issues for the row menu or overlay...
> 
> Do you mean #517? Why do I have to link it 3 times 😁. Ok so it's a dependency that we need to solve first before we can solve the Cell editing (cell editing depends on the these actions).

It can be done separately before or after the edit row. Editing row can be done via an edit button the toolbar for example. These are 2 separate issues.

> 
> > Therefore I do not see the point of `valueParser` or `valueSetter`
> 
> `valueParser` I'm not sure that the parse of the new value can be done inside `renderEditCell`, this components needs to control the row value. The parse step seems to only make sense once the user wants to validate his action.

It can all be done in the renderEditCell. That would mean that users would have to customise the `renderEditCell` for that.
As it might not be that convenient, we can create points where users can hook into. 
Our current overall approach would suggest adding a new handler such as `onCellValueChange`, that would allow to validate the input, or even do a server call...

**Comment by oliviertassinari** [comment]
>  Therefore I do not see the point of `valueParser` or `valueSetter`

`valueParser` Is about converting the value rendered inside `renderEditCell` into a value that can be stored inside the row. `valueSetter` Is about storing the new value inside the row.

It seems that `renderEditCell` should be **stateless** in order to implement all the controls. By making it stateless, we make the format and update steps something that `renderEditCell` can't have full control over hence these two functions would still be required.
